components/style/animation.rs

```diff
@@ -369,56 +365,182 @@ impl PropertyAnimation {
     }
 }
 
-/// Start any new transitions for this node and ensure that any existing transitions
-/// that are cancelled are marked as cancelled in the SharedStyleContext. This is
-/// at the end of calculating style for a single node.
+/// Holds the animation state for a particular element.
 #[cfg(feature = "servo")]
-pub fn update_transitions(
-    context: &SharedStyleContext,
-    new_animations_sender: &Sender<Animation>,
-    opaque_node: OpaqueNode,
-    old_style: &ComputedValues,
-    new_style: &mut Arc<ComputedValues>,
-    expired_transitions: &[PropertyAnimation],
-) {
-    let mut all_running_animations = context.running_animations.write();
-    let previously_running_animations = all_running_animations
-        .remove(&opaque_node)
-        .unwrap_or_else(Vec::new);
-
-    let properties_that_transition = start_transitions_if_applicable(
-        context,
-        new_animations_sender,
-        opaque_node,
-        old_style,
-        new_style,
-        expired_transitions,
-        &previously_running_animations,
-    );
-
-    let mut all_cancelled_animations = context.cancelled_animations.write();
-    let mut cancelled_animations = all_cancelled_animations
-        .remove(&opaque_node)
-        .unwrap_or_else(Vec::new);
-    let mut running_animations = vec![];
-
-    // For every animation that was running before this style change, we cancel it
-    // if the property no longer transitions.
-    for running_animation in previously_running_animations.into_iter() {
-        if let Animation::Transition(_, _, ref property_animation) = running_animation {
-            if !properties_that_transition.contains(property_animation.property_id()) {
-                cancelled_animations.push(running_animation);
-                continue;
+#[derive(Default)]
+pub struct ElementAnimationState {
+    /// The animations running for this element.
+    pub running_animations: Vec<Animation>,
+
+    /// The animations that have finished for this element, but not canceled. These are cleared
+    /// upon triggering a DOM event for each finished animation.
+    pub finished_animations: Vec<Animation>,
+
+    /// The animations that have been cancelled for this element. These are cleared
+    /// upon triggering a DOM event for each cancelled animation.
+    pub cancelled_animations: Vec<Animation>,
+
+    /// New animations created for this element.
+    pub new_animations: Vec<Animation>,
+}
+
+#[cfg(feature = "servo")]
```

Comment:
I wouldn't bother cfg'ing out these fwiw. The less `#[cfg]`d code the better. Makes it harder for me to break servo, and makes it harder for Servo to break Gecko with unused imports and so on.

Reply:
Okay. I've removed these.